gocov/test.go

```diff
@@ -21,12 +21,113 @@
 package main
 
 import (
+	"bytes"
+	"fmt"
 	"io/ioutil"
 	"os"
 	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
 )
 
+func capture(wd string, args []string) ([]byte, error) {
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdin = nil
+	out := &bytes.Buffer{}
+	cmd.Stdout = out
+	cmd.Stderr = out
+	cmd.Dir = wd
+	err := cmd.Run()
+
+	out2 := &bytes.Buffer{}
+	prefix := "warning: no packages being tested depend on "
+	for _, line := range strings.SplitAfter(out.String(), "\n") {
+		if !strings.HasPrefix(line, prefix) {
+			out2.WriteString(line)
+		}
+	}
+	return out2.Bytes(), err
+}
+
+func readDirNames(dirname string) []string {
+	f, err := os.Open(dirname)
+	if err != nil {
+		return nil
+	}
+	names, err := f.Readdirnames(-1)
+	_ = f.Close()
+	return names
+}
+
+// relToGOPATH returns the path relative to $GOPATH/src.
+func relToGOPATH(p string) (string, error) {
+	for _, gopath := range filepath.SplitList(os.Getenv("GOPATH")) {
+		if len(gopath) == 0 {
+			continue
+		}
+		srcRoot := filepath.Join(gopath, "src")
+		// TODO(maruel): Accept case-insensitivity on Windows/OSX, maybe call
+		// filepath.EvalSymlinks().
+		if !strings.HasPrefix(p, srcRoot) {
+			continue
+		}
+		rel, err := filepath.Rel(srcRoot, p)
+		if err != nil {
+			return "", fmt.Errorf("failed to find relative path from %s to %s", srcRoot, p)
+		}
+		return rel, err
+	}
+	return "", fmt.Errorf("failed to find GOPATH relative directory for %s", p)
+}
+
+// goTestDirs returns the list of directories with '*_test.go' files.
```

Comment:
This function doesn't consider build tags. I'm not sure that it's really needed anyway.